### DISCUSSION TITLE
Decouple CLI from API/SDK

### DISCUSSION THREAD

**Discussion by Boeller666** [siderolabs/talos#5594] [Ideas]
## Overview
The CLI is often tightly coupled to other package components, such as the machinery package. Sometimes the code is also orchestrated directly within the CLI command.

## Why decoupling?
I tried to create a Terraform provider for Talos, but found it very difficult to get beyond generating the configuration. The main problem besides finding the right packages is that Talos relies on persisted files and late changes to them, such as configuring the `talosconfig` endpoint or nodes.  
  
Decoupling and creating an SDK-like package that can then be used within the Cobra-based `talosctl` would greatly simplify adoption for other automation tools such as *Terraform* or *Controlplane*.

## Why a provider for Terraform or Controlplane?
I tried to create a Terraform provider instead of using a `local_exec` provisioner or other shell adoptions inside Terraform to easily get the results of the build process inside a Terraform state and to get rid of the dependency to the locally installed talosctl version. That way it would also be easier to track changes and integrate the IaC part into Git and be reliable every time I need to rebuild a cluster.
The same should be true for easy infrastructure creation via Controlplane.

**Comment by frezbo** [comment]
you can take a look at this: https://github.com/siderolabs/pulumi-provider-talos/blob/main/provider/pkg/provider/provider.go

it doesn't require access to any local files

**Comment by frezbo** [comment]
you can also look at this example code [here](https://github.com/siderolabs/talos/blob/master/pkg/machinery/config/types/v1alpha1/generate/example_test.go) from #5461

**Comment by smira** [comment]
First of all, we would love to support any effort to create providers for Talos in Terraform or any other automation tool.

Talos exposes `machinery` package https://pkg.go.dev/github.com/talos-systems/talos/pkg/machinery which can be used to generate and manipulate machine configuration, access Talos APIs, etc.

Not all things are in perfect order, but we are open to improve things based on the feedback. 

Sidero Labs also provides some CAPI tooling which relies on the machinery package, so it can be viewed as examples to use the code:

* https://github.com/siderolabs/cluster-api-bootstrap-provider-talos/ (generating machine config/patching)
* https://github.com/siderolabs/cluster-api-control-plane-provider-talos (API access, checks, control plane scale up/down)
